### PR TITLE
Guided Tours: Avoid collisions with First View

### LINKED ISSUE #6996: Guided Tours / First View: don't collide
Not an issue currently, as no guided tour is enabled in production.

screen shot 2016-07-25 at 14 19 07
## Preliminary thoughts
- Treat First View (FV) displays like guided tours
- That way, a single selector `hasRecentlySeenTours` ("recently" could be defined as a couple of minutes) could be checked in Guided Tours (GT) selectors, so that no new tour is initiated if the former returns `true`.
- That said, there are different UX considerations for FV vs. GT — for instance, it makes sense to show different FV dialogs one after the other for a new user that is discovering Calypso's sections (e.g. lands in `/stats`, sees FV for Stats, switches to `/pages`, sees FV for Pages), but it doesn't make sense to do the same for GT.
- So maybe two separate conditions are to be checked — `isShowingFirstView` and `hasRecentlySeenTours` by GT?

cc @mattsherman

### PR DESCRIPTION
**Pull request by mcsf**
Fixes #6996 

To test:

1. Apply the patch found at the end of this description, just so you can pretend to be a new user.
2. Open Calypso and `/stats`, make sure only Guided Tours appears, not First View
3. Look at the patch again; play with its values — e.g. in `MainTour`, switch the `when` predicate to return `false` instead of `true` — and make sure GT/FV behavior remains sane — i.e., for the `MainTour` example, make sure that now only First View appears.
4. Generally speaking, try to break things. :)

```diff
diff --git a/client/layout/guided-tours/main-tour.js b/client/layout/guided-tours/main-tour.js
index 1e190fd..b37cbf0 100644
--- a/client/layout/guided-tours/main-tour.js
+++ b/client/layout/guided-tours/main-tour.js
@@ -27,7 +27,7 @@ const scrollSidebarToTop = () =>
 	scrollTo( { y: 0, container: getScrollableSidebar() } );
 
 export const MainTour = makeTour(
-	<Tour name="main" version="20160601" path="/" when={ and( isNewUser, isEnabled( 'guided-tours/main' ) ) }>
+	<Tour name="main" version="20160601" path="/" when={ () => true }>
 		<Step name="init" placement="right" className="guided-tours__step-first">
 			<p className="guided-tours__step-text">
 				{
diff --git a/client/state/ui/first-view/selectors.js b/client/state/ui/first-view/selectors.js
index 01b209d..017eed2 100644
--- a/client/state/ui/first-view/selectors.js
+++ b/client/state/ui/first-view/selectors.js
@@ -37,6 +37,7 @@ export const getConfigForCurrentView = createSelector(
 );
 
 export function isUserEligible( state, config ) {
+	return true;
 	const userStartDate = getCurrentUserDate( state );
 
 	// If no start date is defined, all users are eligible
@@ -65,7 +66,7 @@ export function isViewEnabled( state, config ) {
 		return false;
 	}
 
-	const latestFirstViewHistory = findLast( getPreference( state, 'firstViewHistory' ), { view: config.name } );
+	const latestFirstViewHistory = findLast( []||getPreference( state, 'firstViewHistory' ), { view: config.name } );
 	const isViewDisabled = latestFirstViewHistory ? ( !! latestFirstViewHistory.disabled ) : false;
 
 	// If the view is disabled, we want to return false, regardless of state
diff --git a/client/state/ui/guided-tours/selectors.js b/client/state/ui/guided-tours/selectors.js
index cceff9b..5932da9 100644
--- a/client/state/ui/guided-tours/selectors.js
+++ b/client/state/ui/guided-tours/selectors.js
@@ -75,7 +75,7 @@ const getToursFromFeaturesReached = createSelector(
  * recently and in the past.
  */
 const getToursSeen = createSelector(
-	state => uniq( map( getToursHistory( state ), 'tourName' ) ),
+	state => []||uniq( map( getToursHistory( state ), 'tourName' ) ),
 	getToursHistory
 );
```

### CODE REVIEW

**Comment by mcsf** [comment]
gt-over-fv

At the moment, the changes are achieving the PR's goal, which is to prevent **new tours** from popping up:
- on top of First View screens;
- **or** immediately (within 60 seconds) after another tour.

**Comment by lsinger** [comment]
Maybe we also want to do the complement — prevent First View views from appearing when a tour is currently running?

**Comment by mattsherman** [comment]
@mcsf I took a quick look at the `first-view` changes -- makes sense to me (no change in functionality, right? just refactoring?).

**Comment by mattsherman** [comment]
@lsinger That makes sense.

**Comment by mcsf** [comment]
> Maybe we also want to do the complement — prevent First View views from appearing when a tour is currently running?

@lsinger: that was my first instinct, but I realized that that may introduce undesired effects and would be a decision to make with Tanooki — suppose there was a tour that took the user through a couple of screens and that we'd like to _not_ expose the user to sad empty states of said screens, so we'd prefer _not_ to disable FV. Conversely, if we did disable FV during that tour, what would be better for the user after the tour is completed? To show FV for those screens (delivering something of an anticlimactic experience, as the user already has seen what lies beneath FV), or not to show FV, thus resulting in the user having _never_ seen FV for a number of features.

That was the thought process that held me from adding that check. That said, I can see the reasoning for exactly the opposite: maybe we want a first-run tour to take the user through a couple of screens so that they can perform some specific actions, and we don't want FV to stand in the way of that.

Your call, @mattsherman and @lsinger :) — however, can we make that a separate PR, to avoid blockage here?

> makes sense to me (no change in functionality, right? just refactoring?).

Correct!

**Comment by lsinger** [comment]
> can we make that a separate PR, to avoid blockage here?

sounds good to me.

**Comment by mcsf** [comment]
Rebased. Added testing instructions.

**Comment by mattsherman** [comment]
@mcsf I think your comments show that the decision whether or not to prevent FVs from showing while a tour is going on requires further thought. Definitely should be a separate PR if we decide to go that route.

**Comment by mcsf** [comment]
> @mcsf I think your comments show that the decision whether or not to prevent FVs from showing while a tour is going on requires further thought. Definitely should be a separate PR if we decide to go that route. 

Actually, in the process of getting this ready for testing, I realized that, due to a race between GT and FV (and because FV explicitly waits for fresh `preferences`), in practice GT would always start first, despite its checks to not start it FV is showing. In other words, when testing collision avoidance you would get both FV and GT showing at once. See penultimate commit for the fix to that by having FV itself not collide with GT. 

There _could_ be other fixes out there, like having GT also wait for fresh `preferences` or artificially waiting for something else. But this was the quickest win. We can revise.

**Comment by lsinger** [comment]
Note that `client/state/ui/guided-tours/selectors.js` is now (mostly) in `client/state/ui/guided-tours/selectors/index.js`. Assuming that's on purpose @mcsf?

**Comment by lsinger** [comment]
Update: I applied the patch above, but I can't seem to be able to trigger a FirstView. 😕

**Comment by lsinger** [comment]
It seems to be that `getConfigForCurrentView` isn't returning anything non-falsy. The following works though: 

```
	if ( firstViewConfig ) {
		return true;
	}
```

**Comment by mcsf** [comment]
> Note that client/state/ui/guided-tours/selectors.js is now (mostly) in client/state/ui/guided-tours/selectors/index.js. Assuming that's on purpose @mcsf?

Yes; this allows every module importing GT selectors to use the same path: `state/ui/guided-tours/selectors`.

File: client/state/ui/first-view/selectors.js [hunk 1]
```diff
@@ -78,14 +90,13 @@ function routeSetIsInCurrentSection( state, routeSet ) {
 	return some( section.paths, path => startsWith( routeSet.path, path ) );
 }
 
+// TODO: memoize
```
Comment (marekhrabe) [inline client/state/ui/first-view/selectors.js]: are leaving this TODO in so it's not forgotten in future or will you address it now?

**Comment by marekhrabe** [comment]
I was playing with this over the weekend and I wasn't able to trigger both GT and FV at the same time. I was using the patch from description to simulate NUX. Overall it looks very good! 👍

**Comment by marekhrabe** [comment]
> Update: I applied the patch above, but I can't seem to be able to trigger a FirstView (i.e. even with the tour trigger being set to false). 😕

@lsinger with that patch applied, I was able to see FV, right when I quit the tour

File: client/state/ui/first-view/selectors.js (hunk 1 above)
Comment (lsinger) [inline client/state/ui/first-view/selectors.js]: If you want to leave it open, maybe add (@mcsf).

**Comment by lsinger** [comment]
I reset my workspace and reapplied the patch. I must've done something wrong the first time. Behavior looks good now:

- If I go to http://calypso.localhost:3000/stats/day the StatsFV shows. 
- If I go to http://calypso.localhost:3000/ the `main` tour plays. 
- If I come across http://calypso.localhost:3000/stats/day while the `main` tour plays, the StatsFV does NOT show, just as it should. 

LGTM. :shipit: Would be great if you could clear up the `TODO` comment.

File: client/state/ui/first-view/selectors.js (hunk 1 above)
Comment (mcsf) [inline client/state/ui/first-view/selectors.js]: I'll remove it. It's non trivial to optimize properly, I think.

**Pull request merged**